grails-app/services/org/pih/warehouse/inventory/CycleCountService.groovy

```diff
@@ -301,29 +307,82 @@ class CycleCountService {
     List<Map> getRecountFormXls(List<CycleCountDto> cycleCounts) {
         List<Map> data = []
         cycleCounts.each { CycleCountDto cycleCount ->
-            cycleCount.cycleCountItems?.findAll { it.countIndex == 0 }?.each { CycleCountItemDto item ->
-                data << [
-                        "Product Code": item.product.productCode,
-                        "Product Name": item.product.name,
-                        "Lot Number": item.inventoryItem.lotNumber,
-                        "Expiration Date": item.inventoryItem.expirationDate
-                                ? Constants.EXPIRATION_DATE_FORMATTER.format(item.inventoryItem.expirationDate) : "",
-                        "Bin Location": item.binLocation?.name,
-                        "Quantity Counted": item.quantityCounted,
-                        "Difference": item.quantityVariance,
-                        "Counted by": item.assignee,
-                        "Date Counted": item.dateCounted ? Constants.EXPIRATION_DATE_FORMATTER.format(item.dateCounted) : "",
-                        "Quantity Recounted": "",
-                        "Comment": "",
-                        "Recounted By": "",
-                        "Date Recounted": ""
-                ]
+            data.addAll(getRecountAsXlsMap(cycleCount))
+        }
+
+        return data
+    }
+
+    /**
+     * Converts a CycleCountDto to a list of rows/maps. For use when exporting to XLS.
+     */
+    private List<Map> getRecountAsXlsMap(CycleCountDto cycleCount) {
+        int currentCountIndex = cycleCount.maxCountIndex
+
+        // Build a map to make it easier to group the count and recount items into one row in the XLS.
+        // The outer map is keyed on [product code + lot + bin]. The inner map is keyed on count index.
+        MultiKeyMap<String, Map<Integer, CycleCountItemDto>> countItemsMap = [:]
```

Comment:
I quite like MultiKeyMap and MultiKey. I find them very useful as a way to avoid needing to make nested maps (though in this rare case I *also* have a nested map for convenience).

https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/map/MultiKeyMap.html